### PR TITLE
Iterators and ranges library

### PR DESCRIPTION
**Pull request by renehiemstra**
First implementation of iterators, ranges, range adaptors, and range combiners. 

A range is defined as something that implements the '__for' metamethod
An iterator implements a  'getfirst', 'getnext', and 'islast' macro. the '__for' metamethod is generated automatically.

Couple of notes:
* collect now works on containers that implement 'push', 'pop', and 'size'.
* the range combiners are ranges, but not implemented as iterators, so they don't implement a 'getfirst', 'getnext', and 'islast' macro. Maybe we want to change that.
* Maybe it's possible to change the 'getfirst', 'getnext', and 'islast' macros to terra functions. Then it would be easier to define the iterator interface.

### CODE REVIEW

File: range.t [hunk 1]
```diff
@@ -0,0 +1,718 @@
...
+                var state_1, value_1 = iter._1:getfirst()
+                var state_2, value_2 = iter._2:getfirst()
+                while not (iter._0:islast(state_0, value_0) or iter._1:islast(state_1, value_1) or iter._2:islast(state_2, value_2)) do
+                    [body(value_0, value_1, value_2)]
+                    value_0 = iter._0:getnext(state_0)
+                    value_1 = iter._1:getnext(state_1)
+                    value_2 = iter._2:getnext(state_2)
+                end
+            end
+        end
```
Comment (tpkessler) [inline range.t:670]: Add `error()` if `D` is larger than 3?

File: range.t [hunk 2]
```diff
@@ -0,0 +1,718 @@
...
+    filter = filter,
+    take = take,
+    drop = drop, 
+    take_while = take_while,
+    drop_while = drop_while,
+    enumerate = enumerate,
+    join = join,
+    product = product,
+    zip = zip,
+    develop
```
Comment (tpkessler) [inline range.t:717]: `develop = develop`?

File: test_lambdas.t [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import "terratest/terratest"
+local lambdas = require("lambdas")
+
+testenv "lambda's" do
+
+    testset "no captures" do
+        terracode
+            var p = lambdas.lambda([terra(i : int) return i * i end])
```
Comment (tpkessler) [inline test_lambdas.t:8]: Do we really need `[expr]` for the lambdas?

Comment (renehiemstra) [reply test_lambdas.t:8]: yes. terra functions need to be inside lua code.

File: range.t (hunk 2 above)
Comment (renehiemstra) [reply range.t]: fixed.

File: range.t (hunk 1 above)
Comment (renehiemstra) [reply range.t:680]: added error handling to zip and product ranges.

**Pull request merged**